File: permutations/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        nums_set = set(nums)
        ret = []
        
        def rec(num, nums_set, retLi):
            nums_set_dup = set(nums_set)
            nums_set_dup.remove(num)
            retLi += [num]
            if len(nums_set_dup) == 0:
                ret.append(retLi)
            for numi in nums_set_dup:
                rec(numi, nums_set_dup, list(retLi))
            
        
        for num in nums:
            rec(num, nums_set, [])
        return ret

File: permutations/permutations.java
class Solution {
    List<List<Integer>> ret = new ArrayList<List<Integer>>();
    int[] nums;
    
    public List<List<Integer>> permute(int[] nums) {
        this.nums = nums;
        backtrack(new ArrayList<Integer>());
        return ret;
    }
    
    private void backtrack(List<Integer> curr) {
        if(curr.size() == nums.length) {
            ret.add(new ArrayList<Integer>(curr));
        }
        var set = new HashSet<Integer>(curr);
        for(int i=0; i<nums.length; i++) {
            if(!set.contains(nums[i])) {
                curr.add(nums[i]);
                backtrack(curr);
                curr.remove(curr.size()-1);
            }
        }
    }
}